include/ugfx.h

```diff
@@ -0,0 +1,601 @@
+
+#define ugfx_load_light(slot, light, index) (\
+    __ugfx_opcode(UGFX_OP_LOAD_LIGHT) | \
+    __ugfx_mask_shift((index), 0x7, 35) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(light), 0x1FFFFFF, 0))
+
+#define ugfx_link_commands(slot, commands, length, flags) (\
+    __ugfx_opcode(UGFX_OP_LINK_COMMANDS) | \
+    __ugfx_mask_shift((length), 0x1FFFFF, 35) | \
+    __ugfx_mask_shift((flags), 0x7, 32) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(commands), 0x1FFFFFF, 0))
+
+#define ugfx_push_commands(slot, commands, length) ugfx_link_commands(slot, commands, length, UGFX_LINK_PUSH)
+#define ugfx_load_commands(slot, commands, length) ugfx_link_commands(slot, commands, length, UGFX_LINK_LOAD)
+
+#define ugfx_set_cull_mode(mode) (\
+    __ugfx_opcode(UGFX_OP_SET_CULL_MODE) | \
+    __ugfx_mask_shift((mode), 0x3, 32))
+
+#define ugfx_set_geometry_mode(mode) (\
+    __ugfx_opcode(UGFX_OP_SET_GEOMETRY_MODE) | \
+    __ugfx_mask_shift((mode), 0xFF, 32))
+
+#define ugfx_set_persp_norm(scale) (\
+    __ugfx_opcode(UGFX_OP_SET_PERSP_NORM) | \
+    __ugfx_mask_shift((scale), 0xFFFF, 32))
+
+#define ugfx_set_texture_settings(scale_s, scale_t, level, tile) (\
+    __ugfx_opcode(UGFX_OP_SET_TEXTURE_SETTINGS) | \
+    __ugfx_mask_shift((level), 0x7, 35) | \
+    __ugfx_mask_shift((tile), 0x7, 32) | \
+    __ugfx_mask_shift((scale_s), 0xFFFF, 16) | \
+    __ugfx_mask_shift((scale_t), 0xFFFF, 0))
+
+#define ugfx_set_num_lights(num) (\
+    __ugfx_opcode(UGFX_OP_SET_NUM_LIGHTS) | \
+    __ugfx_mask_shift((num), 0x7, 35))
+
+#define ugfx_set_clip_ratio(ratio) (\
+    __ugfx_opcode(UGFX_OP_SET_CLIP_RATIO) | \
+    __ugfx_mask_shift((ratio), 0xFFFF, 32))
+
+#define ugfx_set_address_slot(slot, address) (\
+    __ugfx_opcode(UGFX_OP_SET_ADDRESS_SLOT) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(address), 0x1FFFFFF, 0))
+
+#define ugfx_draw_triangle(v0, v1, v2) ( \
+    __ugfx_opcode(UGFX_OP_DRAW_TRIANGLE) | \
+    __ugfx_mask_shift((v0), 0x3F, 49) | \
+    __ugfx_mask_shift((v1), 0x3F, 43) | \
+    __ugfx_mask_shift((v2), 0x3F, 37))
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct ugfx_vertex_t {
+    int16_t x;
+    int16_t y;
+    int16_t z;
+    int16_t padding;
+    int16_t s;
+    int16_t t;
+
+    union {
+        uint32_t rgba;
+
+        struct {
+          uint8_t r;
+          uint8_t g;
+          uint8_t b;
+          uint8_t a;
+        } color;
+
+        struct {
+          int8_t x;
+          int8_t y;
+          int8_t z;
+          uint8_t a;
+        } normal;
+    } attr;
+} __attribute__((aligned(16))) ugfx_vertex_t;
+
+typedef struct ugfx_matrix_t {
+    int16_t integer[4][4];
+    uint16_t fraction[4][4];
+} __attribute__((aligned(16))) ugfx_matrix_t;
+
+void ugfx_matrix_from_column_major(ugfx_matrix_t *dest, const float *source);
+void ugfx_matrix_from_row_major(ugfx_matrix_t *dest, const float *source);
```

Comment:
I have played around with [cglm](https://github.com/recp/cglm) before, which works on float matrices. I'm not sure if using the FPU is worth the convenience vs the performance, though. Some profiling should be done to assess this, I think. Either way, we should probably offer our own set helpers for creating matrices. I don't think we could just incorporate a library like cglm into libdragon...?

I agree that we should stick to one matrix format. My pick would be column major, because that's the format used by the microcode anyway. Matrix multiplications just work out to be much easier and more efficient that way. (Libultra uses row major by the way and I wonder what made them decide on that format)

Reply:
Nice I didn't know that library. I wouldn't incorporate it verbatim though, it's too much stuff that we don't need for 3d graphics, plus we don't need the SSE versions, etc. But we can borrow some nice ideas I saw on the APIs for "modern C".

Column major is fine by me.